simularium/simularium-viewer
# Feature/orthographic camera

Problem
=======
Resolves #305: Users would like the option to switch between perspective and orthographic camera views.

Solution
========
- Add `perspectiveCamera` and `orthographicCamera` properties to `VisGeometry`
- Add the public method `setCameraType` to switch which of these is active in the `camera` property
- Add some private methods for conveniently dealing with switching between cameras (see below)
- Update some other methods to consider orthographic cameras (`dolly`, `storeCamera`, `loadCamera`)
- Add fov to tweakpane and make camera saving/loading/importing/exporting from there more robust
- Do some formatting and cleanup along the way

Orthographic cameras render all objects in the scene at a consistent size regardless of distance, while the size of objects in perspective cameras decreases with size. Keeping sizes of objects looking reasonably consistent when switching views requires picking a distance at which both cameras will render the same object at the same size. After some testing, I found that picking the default distance from the camera to the origin kept things looking reasonable.

Additionally, `OrbitControls` handles zoom differently between perspective and orthographic cameras. To zoom a perspective camera, it physically moves the camera closer to or farther from the target. To zoom an orthographic camera, it doesn't touch the distance from the camera to the target and instead modifies the camera's `zoom` property. A lot of the technical challenge here was a result of having to reconcile these two zoom methods sharing one controller.

Given the notes above, I added a few private methods to `VisGeometry`:
- `getDefaultOrbitRadius`: Derives the default/initial distance between the camera and the target from `cameraDefault`. 98% of the time this ought to be equal to `DEFAULT_CAMERA_Z_POSITION` (120), but trajectories with non-standard units or camera positions will expect the app to notice and keep things consistent.
- `updateOrthographicFrustum`: The default "size" of the orthographic camera varies with the aspect of the frame, the FOV of the perspective camera (in other words, how fast objects get smaller as they get farther from the camera), and the default orbit radius. This method updates that if anything changes.
- `updateControlsZoomBounds`: `OrbitControls` bounds how much the camera can zoom in/out by distance for perspective cameras but by zoom level for orthographic cameras. These could all be set as constants, but the relationship between them will vary if the default orbit radius changes. So this method is here to keep things consistent should that change.

I also had to make some pretty significant changes to the `dolly` method to handle both ways of zooming a camera.

### Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. `npm start`
2. Click new "Camera mode" button to switch between perspective and orthographic views
3. Move/zoom the view while switching to verify that position and zoom are preserved reasonably between modes
4. Verify zooming in/out with +/- buttons works in orthographic mode
5. Open tweakpane with Ctrl+Alt+1, verify that setting/saving/loading camera state from there works too

Screenshots:
-----------------------
![image](https://github.com/simularium/simularium-viewer/assets/53030214/11bdb915-d2d7-4a3e-b705-36a87a258dad)
![image](https://github.com/simularium/simularium-viewer/assets/53030214/226bda00-1c08-43e6-afb7-f8dbef8f7277)
